packet03-redirecting/README.org

```diff
@@ -5,49 +5,277 @@
 Now that you have come this far, you know how to parse packet data, and how
 to modify packets. These are two of the main components of a packet
 processing system, but there is one additional component that is missing:
-How to redirect packets and transmit them back out onto the network.
-
-This lesson will cover this aspect of packet processing. Unfortunately, we
-have not had time to complete it yet. So if you've made it this far,
-congratulations, you are at the end of the tutorial for now!
-
-If you do feel up to a challenge, the headings below should give you a hint
-about what to do in this lesson; but you will have to find guidance and
-examples elsewhere. For instance, you could take a look at the [[https://elixir.bootlin.com/linux/latest/source/samples/bpf][examples in
-the kernel source tree]].
-
-Bonus points for sending a pull request with extensions of this tutorial
-document when you're done :)
+how to redirect packets and transmit them back out onto the network.
+This lesson will cover this aspect of packet processing.
 
 * Table of Contents                                                     :TOC:
+- [[#what-you-will-learn-in-this-lesson][What you will learn in this lesson]]
+  - [[#sending-packets-back-to-the-interface-they-came-from][Sending packets back to the interface they came from]]
+  - [[#redirecting-packets-to-other-interfaces][Redirecting packets to other interfaces]]
+  - [[#building-a-router-using-a-kernel-helper][Building a router using a kernel helper]]
 - [[#assignments][Assignments]]
   - [[#assignment-1-send-packets-back-where-they-came-from][Assignment 1: Send packets back where they came from]]
   - [[#assignment-2-redirect-packets-between-two-interfaces][Assignment 2: Redirect packets between two interfaces]]
   - [[#assignment-3-extend-to-a-bidirectional-router][Assignment 3: Extend to a bidirectional router]]
   - [[#assignment-4-use-the-bpf-helper-for-routing][Assignment 4: Use the BPF helper for routing]]
 
+* What you will learn in this lesson
+
+** Sending packets back to the interface they came from
+
+The =XDP_TX= return value can be used to send the packet back from the same
+interface it came from.  This functionality can be used to implement load
+balancers, to send simple ICMP replies, etc.  We will use this functionality in
+the Assignment 1 to implement a simple ICMP echo server.
+
+Note that in all our assignments below we are using the =skb= (i.e.,
+=xdpgeneric=) mode to load programs.  We are doing this because =veth= devices
+won't deliver redirected/retransmitted XDP frames unless there is an XDP
+program attached to the receiving side of the target =veth= interface. Physical
+hardware will likely behave the same. XDP maintainers are currently working on
+fixing this behaviour upstream.
```

Comment:
I think we should still use the native mode in later assignments; it may be a while before this is fixed upstream, and in the meantime it's important to show a user how they are going to get things working on real hardware.

So explain that loading an XDP program on the other interface is needed, and make it part of assignment 3 and 4 to actually do this.

Reply:
OK, sure. I also will add a link to this talk: https://www.netdevconf.org/0x13/session.html?talk-veth-xdp